components/create_comment/create_comment.jsx

```diff
@@ -255,6 +256,10 @@ export default class CreateComment extends React.PureComponent {
         }
 
         await this.doSubmit(e);
+
+        const fasterThanHumanWillClick = 150;
```

Comment:
@CoolTomatos, can you elaborate on what moving this block fixes? I've seen some recent fixes around this block and want to understand this better. In particular, I note that we now change the focus in cases where `doSubmit` might have exited early.

Reply:
@lieut-data 
My first try was to add
```
if (Utils.isFeatureEnabled(PreReleaseFeatures.MARKDOWN_PREVIEW)) {
    if (this.refs.textbox.state.preview) {
        this.refs.textbox.hidePreview();
    }
}
```
right before
```
const fasterThanHumanWillClick = 150;
const forceFocus = (Date.now() - this.lastBlurAt < fasterThanHumanWillClick);
this.focusTextbox(forceFocus);
```
at its original place. But the thing is that the focus won't go back to the editor.
I think the reason is that after `hidePreview()`, though we exit preview mode, the editor is still hidden.
So  `focusTextbox()` doesn't work.
So I moved it out to make sure the moment when this piece of code is executed, the editor has been rendered as visible.
But you are right about changing the focus even when `doSubmit()` exits early.
Maybe put this block inside the `try` block after `await this.props.onSubmit()`?

Reply:
Thanks for your patience on this one, @CoolTomatos!

Unfortunately, the proposed fix was something @saturninoabril recently tackled and discovered the need to focus /before/ the await call: https://github.com/delftswa2018/mattermost-webapp/commit/b2961a508133dedb24b1c1e22fad0b751b659611#diff-59fc657c354b43771e2e46f5089c1260. Basically, anytime `await this.props.onSubmit` took too long, the focus timing logic broke the linked issue.

I'm worried that we're all going in circles on this component. I think the fundamental problem is the use of internal state on the textbox that is being manipulated by function calls instead of the normal React props flow. In an ideal world, `showPreview` would be a prop, with the state for this instance of the component stored in the [rhs view reducer](https://github.com/mattermost/mattermost-webapp/blob/master/reducers/views/rhs.js). @jwilander, @hmhealey, is this a sensible usage of same?

Obviously, I don't expect you, @CoolTomatos, to have to tackle all of this... but if you're up for the challenge, I'd love to see this component stop breaking mysteriously each release :D

Reply:
(I'll also say that there might yet still be a solution to this problem that preserves all the previous semantics and doesn't require a refactor, but just want to loop in everyone who's been involved with issues in this component first to get feedback.)

Reply:
Sorry, missed that mention. Yeah preferably this would be handled by using redux state like @lieut-data mentioned. It's a bit more work, but basically it would be add a reducer for `showCommentPreview`, pass it in as a prop to the component and use it to display whether or not the preview is showing and then use an action to toggle the state. Then we could handle focus in `componentDidUpdate`